### ISSUE TITLE
over with lensPath turns arrays into objects

### ISSUE THREAD

**Issue by RedNifre** [ramda/ramda#1922]
Modifying nested objects/arrays with a lensPath seems to turn arrays into objects.

E.g. over(lensProp([0,0]), identity, [['x']]) returns {'0': {'0': 'x'}} instead of [['x']]

**Comment by buzzdecafe** [comment]
I think you want `lensIndex` here. 

`over(lensIndex(0), identity, [['x']]) //=> [['x']]`

Also, `lensProp` expects a String, so the fact that it "works" at all when passed `[0, 0]` is ... disappointing

**Comment by RedNifre** [comment]
No, I don't want to use lensIndex here because as you said it only goes down one level.

Consider `var thing = {foo: [{bar:1},{bar:9}]}`

Why should this not work?: `over(lensPath(["foo", 1, "bar"]), add(1), var) // => {foo: [{bar:1},{bar:10}]}`

I find that to be way more intuitive than `over(pipe(lensProp("foo"), lensIndex(1), lensProp("bar")), add(1), var)`

**Comment by mjgpy3** [comment]
@RedNifre You should be able to compose the index lenses, e.g.

```
over(compose(lensIndex(0), lensIndex(0)), identity, [['x']]) 
```
returns `[['x']]`

**Comment by buzzdecafe** [comment]
This is a related discussion: https://github.com/ramda/ramda/issues/1332

**Comment by CrossEye** [comment]
> This is a related discussion: #1332

There was also discussion in #773, and #1839.  In the latter I proposed that we replace the current `lensPath` with something equivalent to 

```js
const lensPath = compose(
  apply(compose),
  map(ifElse(is(Number), lensIndex, lensProp))
);
```

And this would solve the problem expressed by @RedNifre.  But no one responded to that suggestion.  What do people think?  @buzzdecafe?  @davidchambers?  @asaf-romano?  @scott-christopher?  Others?

Before someone dings me on the signature, I would think it appropriate to create a union type such as 

```hs
PrimitiveKey = String | Int
```

and use it for this function:

```hs
lensPath :: [PrimitiveKey] -> Lens
```

**Comment by scott-christopher** [comment]
>  In the latter I proposed that we replace the current lensPath with something equivalent to [...]

Seems sane to me. I've raised #1924 to capture and discuss this approach. Alternatively the implementation of `assocPath` could be updated to support numeric keys which would flow on down to address this issue too.

**Comment by tolmasky** [comment]
I definitely want this behavior.

**Comment by scott-christopher** [comment]
FYI: This is now supported since #1926 landed in master (still yet to be released though).

```js
> R.over(R.lensPath([0, 0]), R.identity, [['x']])
[ [ 'x' ] ]
```

Closing this issue for now.

**Comment by brennancheung** [comment]
Any way we can get a tag made that includes this?  Need something to target with `npm` besides `master`.

**Comment by scott-christopher** [comment]
@brennancheung I have just published v0.23.0
